common/k8s/src/middleware/metadata.rs

```diff
@@ -103,7 +228,107 @@ impl K8sMetadata {
             ListParams::default()
         };
 
-        let watched = StreamBackoff::new(watcher(api, params), ExponentialBackoff::default());
+        let watcher = Box::pin(StreamBackoff::new(
+            watcher(api, params),
+            ExponentialBackoff::default(),
+        ));
+
+        let (deletion_ack_recv, deletion_state) = {
+            let state = self.state.lock().map_err(|_| Error::K8sMiddlewareState)?;
+            // Take the deletion ack stream, copy the deletion state stream
+            (self.deletion_ack_recv.clone(), state.deletion_state.clone())
+        };
+
+        let watched = stream::unfold(
+            (watcher, deletion_state, deletion_ack_recv),
+            move |(mut watcher, deletion_state, mut deletion_ack_recv)| async move {
+                // Check if we have a delete acknowledgement
```

Comment:
Trying to accomplish both deletion ack handling and watcher event intercepting in the same stream will cause them to be in a weird lockstep e.g. if the stream blocks down below waiting for a watcher event, no delete events can be processed.

Could we instead turn these into two separate streams and then merge them together? That should be easier to understand and work quite a bit better.

Finally, in my own testing it didn't seem like this section of code was ever being run; I didn't see `handle_pod` called for any delete events despite the log file being deleted.